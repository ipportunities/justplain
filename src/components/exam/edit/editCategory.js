import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { EditConstraint } from './editConstraint'
import apiCall from '../../api'
import { useTranslation } from '../../translate'
import Saved from "../../intervention/saved";
import { setSavingStatus } from "../../../actions";

export const EditExamTypeCategory = ({ examTypeId, category, examTypeCategory, setExamTypeCategory, updateQuestionsTotal, updateMaxPointRange }) => {
  const [enabled, setEnabled] = React.useState(false)
  const [saving, setSaving] = React.useState(false)
  const [saved, setSaved] = React.useState(false)
  const [id, setId] = React.useState(-1)
  const [amountOfQuestions, setAmountOfQuestions] = React.useState({})
  const dispatch = useDispatch();

  const t = useTranslation()
  const auth = useSelector(state => state.auth)

  React.useEffect(() => {
    setEnabled(examTypeCategory !== undefined)
    let newAmountOfQuestions = {}
    if (examTypeCategory !== undefined) {
      category.subjects.forEach(s => {
        newAmountOfQuestions[s] = {}
        category.difficulties.forEach(d => {
          newAmountOfQuestions[s][d] = examTypeCategory.constraints.find(c => c.subject === s && c.difficulty === d)
            ? +examTypeCategory.constraints.find(c => c.subject === s && c.difficulty === d).amountOfQuestions
            : 0
        })
      })
      setId(examTypeCategory.id)
    } else {
      let difficulties = {}
      category.difficulties.forEach(d => difficulties[d] = 0)
      category.subjects.forEach(s => newAmountOfQuestions[s] = difficulties)
    }
    setAmountOfQuestions(newAmountOfQuestions)
  }, [examTypeCategory, category])

  const switchExamTypeCategoryEnabled = () => {
    const newEnabled = !enabled
    if (newEnabled) {
      apiCall({
        action: 'add_exam_category',
        token: auth.token,
        data: { examTypeId: examTypeId, categoryId: category.id, amountOfQuestions: 0 }
      }).then(resp => {
        setExamTypeCategory(resp.examTypeCategory)
        setId(resp.examTypeCategory.id)
      })
    } else {
      apiCall({
        action: 'remove_exam_category',
        token: auth.token,
        data: { id: id }
      }).then(() => {
        setId(-1)
      })
    }
    setEnabled(newEnabled)
  }

  const saveCategory = () => {
    if (enabled) {
      let constraints = []
      Object.entries(amountOfQuestions).forEach(([subject, a]) => {
        Object.entries(a).forEach(([difficulty, amount]) => {
          constraints.push({
            subject: subject,
            difficulty: difficulty,
            amountOfQuestions: amount
          })
        })
      })
      apiCall({
        action: 'update_exam_category',
        token: auth.token,
        data: {
          examTypeId: +examTypeCategory.examTypeId,
          categoryId: +examTypeCategory.categoryId,
          amountOfQuestions: 0,
          constraints: constraints
        }
      }).then(resp => {
        dispatch(setSavingStatus("saved"));
      })
    }
  }

  React.useEffect(() => {
    dispatch(setSavingStatus("not_saved"));
    saveCategory()
    updateQuestionsTotal(Object.keys(amountOfQuestions).reduce(
      (total, s) => total + Object.keys(amountOfQuestions[s]).reduce(
        (subTotal, d) => subTotal + parseFloat(amountOfQuestions[s][d] || 0), 0
      ), 0)
    )
  }, [amountOfQuestions])

  const updateAmountOfQuestions = (subject, difficulty, amount) => {
    setAmountOfQuestions({
      ...amountOfQuestions,
      [subject]: {
        ...amountOfQuestions[subject],
        [difficulty]: +amount
      }
    })
  }

  return (
    <>
      <div className="row">
        <div className="col-10 pt-1">
          <input type="checkbox" className="form-check-input" checked={ enabled }
                 onChange={ switchExamTypeCategoryEnabled } id={ `examTypeCategoryEnabled-${ category.id }` } />
          <label className="form-check-label"
                 htmlFor={ `examTypeCategoryEnabled-${ category.id }` }>{ category.name }</label>
        </div>
        <div className="col-2 pt-1 text-right align-self-center">
          { enabled ?
            saving ?
              <div className="spinner-border text-primary" />
              : saved ?
              <div className="fa fa-2x fa-check text-success" />
              : <div />
            : <div />
          }
        </div>
      </div>
      { enabled && <>
        <table className="table">
          <thead>
          <tr>
            <th />
            { category.difficulties.map(d => <th
              key={ d }>{ d.charAt(0).toUpperCase() + d.slice(1).toLowerCase() }</th>) }
            <th className="border-left">{ t('Total') }</th>
          </tr>
          </thead>
          <tbody>
          { category.subjects.map(s => <tr className="border-top" key={ s }>
            <th>{ s === '' ? 'No Subject' : s }</th>
            { category.difficulties.map(d => <td key={ `${ d }-${ s }` }>
              <EditConstraint categoryId={ category.id } difficulty={ d } subject={ s } amount={ amountOfQuestions[s][d] }
                              setAmount={ (amount) => updateAmountOfQuestions(s, d, amount) }
                              updateMaxPointRange={ updateMaxPointRange } />
            </td>) }
            <td className="border-left">
              { Object.keys(amountOfQuestions[s]).reduce((total, i) => total + parseFloat(amountOfQuestions[s][i] || 0), 0) }
            </td>
          </tr>) }
          <tr className="border-top">
            <th>
              { t('Total') }
            </th>
            { category.difficulties.map(d => <td key={ `${ d }-total` }>{
              `${ Object.keys(amountOfQuestions).reduce((total, i) => total + parseFloat(amountOfQuestions[i][d] || 0), 0) }`
            }</td>) }
            <td className="border-left">{
              Object.keys(amountOfQuestions).reduce(
                (total, s) => total + Object.keys(amountOfQuestions[s]).reduce(
                  (subTotal, d) => subTotal + parseFloat(amountOfQuestions[s][d] || 0), 0
                ), 0)
            }</td>
          </tr>
          </tbody>
        </table>
      </> }
    </>
  )
}
