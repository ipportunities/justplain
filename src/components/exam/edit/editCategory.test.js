import React from 'react'
import { shallow } from 'enzyme'
import { EditExamTypeCategory } from './editCategory'

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockReturnValue({ auth: { token: 'a' } })
}))

const categoryMock = {
  id: 1,
  name: 'Category A',
  amountOfQuestions: 15,
  difficulties: ['EASY', 'HARD'],
  subjects: ['Theoretical', 'Practical']
}
const examTypeCategoryMock = {
  amountOfQuestions: 15,
  examTypeId: 5,
  categoryId: 1,
  constraints: [
    { id: 8, amountOfQuestions: 10, difficulty: 'EASY', subject: '' },
    { id: 9, amountOfQuestions: 5, difficulty: '', subject: 'Theoretical' }
  ]
}

describe('Exam Type Category Edit', () => {
  it('should render correctly', () => {
    const component = shallow(<EditExamTypeCategory category={ categoryMock } examTypeCategory={ examTypeCategoryMock } />)
    expect(component).toMatchSnapshot()
  })
})
