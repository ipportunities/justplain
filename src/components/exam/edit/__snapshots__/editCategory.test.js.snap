// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exam Type Category Edit should render correctly 1`] = `
<div
  className="col-6 offset-3 p-2"
>
  <div
    className="row"
  >
    <div
      className="col pt-1"
    >
      <input
        checked={true}
        className="form-check-input"
        id="examTypeCategoryEnabled-1"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="form-check-label"
        htmlFor="examTypeCategoryEnabled-1"
      >
        Category A
      </label>
    </div>
  </div>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th />
        <th
          key="EASY"
        >
          Easy
        </th>
        <th
          key="HARD"
        >
          Hard
        </th>
        <th
          className="border-left"
        >
          Total
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="Theoretical"
      >
        <th>
          Theoretical
        </th>
        <td
          key="EASY-Theoretical"
        >
          <EditConstraint
            amount={0}
            categoryId={1}
            difficulty="EASY"
            setAmount={[Function]}
            subject="Theoretical"
          />
        </td>
        <td
          key="HARD-Theoretical"
        >
          <EditConstraint
            amount={0}
            categoryId={1}
            difficulty="HARD"
            setAmount={[Function]}
            subject="Theoretical"
          />
        </td>
        <td
          className="border-left"
        >
          0
        </td>
      </tr>
      <tr
        key="Practical"
      >
        <th>
          Practical
        </th>
        <td
          key="EASY-Practical"
        >
          <EditConstraint
            amount={0}
            categoryId={1}
            difficulty="EASY"
            setAmount={[Function]}
            subject="Practical"
          />
        </td>
        <td
          key="HARD-Practical"
        >
          <EditConstraint
            amount={0}
            categoryId={1}
            difficulty="HARD"
            setAmount={[Function]}
            subject="Practical"
          />
        </td>
        <td
          className="border-left"
        >
          0
        </td>
      </tr>
      <tr
        className="border-top"
      >
        <th>
          Total
        </th>
        <td
          key="EASY-total"
        >
          0
        </td>
        <td
          key="HARD-total"
        >
          0
        </td>
        <td
          className="border-left"
        >
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
